Portworx framework automation phase - 1 completion

jitupawar22 opened the pull request:
This is for completion of portworx framework automation necessary to verify new px framework builds.

Commit:
DCOS portworx framework automation phase - 1 complition
Added rtery mechanism in pytest SDK around 'package repo list' and 'package repo add'

Commit:
Implemented automation for dcos px secrets functionality
Implemented dcos_utils library
Add required functions in px_utils library
Introduced time tunning for some tests those were failing due to fast operations.

piyush-nimbalkar commented on frameworks/portworx/tests/dcos_utils.py:
```diff
@@ -0,0 +1,87 @@
+# This file contains dcos spacific utilities needed for portworx framework. 
```
nit: *specific

piyush-nimbalkar commented on frameworks/portworx/tests/dcos_utils.py:
```diff
@@ -0,0 +1,87 @@
+
+log = logging.getLogger(__name__)
+
+def install_enterprises_cli():
```
nit: *enterprise_cli
We should install the enterprise cli in the job. It should be assumed in the tests that the enterprise cli exists.

piyush-nimbalkar commented on frameworks/portworx/tests/dcos_utils.py:
```diff
@@ -0,0 +1,87 @@
+# command: dcos security org groups create pxservice
+def create_dcos_security_group(group_name):
+    cmd = "security org groups create " + group_name
+    ret_val, std_out, std_err = sdk_cmd.run_raw_cli(cmd)
```
We should have a info or at least debug statement after every successful operation in the tests, so that we know what stage has passed. It will much easier to understand the logs.

jitupawar22 replied:
The pytest SDK in dcos-commons already provides this type of logging.
Below is log snippet from one of test job run-
------------- snippet begins ----------------------------------
======= END: test_update_px_service_py::test_update_node_count
==========
======= START: test_update_px_service_py::test_update_px_image
------------------------------------ snippet end --------------------------

The log also includes PASS/FAIL/ERROR info for each test along with this.

jitupawar22 replied:
The pytest SDK in dcos-commons already provides this type of logging.
Below is log snippet from one of test job run-
------------- snippet begins ----------------------------------
======= END: test_update_px_service_py::test_update_node_count
==========
======= START: test_update_px_service_py::test_update_px_image
------------------------------------ snippet end --------------------------

The log also includes PASS/FAIL/ERROR info for each test along with this.

jitupawar22 replied:
The pytest SDK in dcos-commons already provides this type of logging.
Below is log snippet from one of test job run-
======= END: test_update_px_service_py::test_update_node_count

======= START: test_update_px_service_py::test_update_px_image

The log also includes PASS/FAIL/ERROR info for each test along with this.

piyush-nimbalkar replied:
I mean logs at operation level. In case it gets stuck and keeps running for a long time, we will know where it failed.

piyush-nimbalkar commented on frameworks/portworx/tests/px_utils.py:
```diff
@@ -143,14 +147,27 @@ def px_service_suspend_resume(pod_name):
-    _, vol_size, std_err = sdk_cmd.run_raw_cli(cmd)
+
+    count = 5
+    while count > 0:
```
Decrement count, else this could go in an infinite loop.

jitupawar22 replied:
Good catch. thanks.

piyush-nimbalkar commented on frameworks/portworx/tests/px_utils.py:
```diff
@@ -143,14 +147,27 @@ def px_service_suspend_resume(pod_name):
+    cmd = "portworx volume list | jq '.[] | select(.locator.name == \"" + vol_name +"\") | .spec.encrypted'"
+    _, vol_type, std_err = sdk_cmd.run_raw_cli(cmd)
+
+    return vol_type
```
This will return a string "true" or null if not present. Should we return a boolean "True" or "False" instead?

piyush-nimbalkar commented on frameworks/portworx/tests/px_utils.py:
```diff
@@ -178,3 +195,35 @@ def px_update_volume_size(pod_name, vol_name, new_size):
         raise
+
+# command: /opt/pwx/bin/pxctl secrets dcos login --username px_user1 --password px_user1 --base-path pwx/secrets
+def px_dcos_login(pod_name, user_name, user_password, base_path):
```
Ideally you shouldn't be doing this. Selecting the secret store during install or update should make px use dcos secrets as the default secret store. The login command is a hack and we will soon remove it.

jitupawar22 replied:
So keeping it as it is for now.

piyush-nimbalkar replied:
We should not do this for testing. Customers don't take this path. It is a shortcut that we take for our testing, it is not a real life scenario. We should try to test whatever could happen in the field.

jitupawar22 replied:
Agreed!

piyush-nimbalkar commented on frameworks/portworx/tests/px_utils.py:
```diff
@@ -178,3 +195,35 @@ def px_update_volume_size(pod_name, vol_name, new_size):
+        raise
+
+# command: /opt/pwx/bin/pxctl volume create --secure --secret_key px_skey  enc_vol1
+def px_create_encrepted_volume(pod_name, vol_name, secret_key):
```
nit: *encrypted

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -149,12 +136,18 @@ def test_start_plan():
     pod_name = pod_list[1]
+
     px_utils.px_create_volume(pod_name, "px_dcos_vol_1", 5)
+    sleep(30) # Wait before immediatly calling vol size, it is observed that dcos needs few seconds to refresh
```
30 seconds is a very long wait. Ideally it should not take that long. Have you tried 5 seconds?

jitupawar22 replied:
I already tried for 5, 10, 15 seconds but observed failures sometimes. So increased it to 30 to be on safer side. There might be probable bug?

piyush-nimbalkar replied:
Did you try to create volume directly from pxctl? How long did it take there?

jitupawar22 replied:
This particular sleep is before calling the command for getting volume size after creating volume. I observed calling volume create or getting volume size directly from the pxctl does  not take that long. its less than 5 seconds.

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -190,6 +185,72 @@ def test_vol_delete():
+    if dcos_utils.check_secret_present(config.PX_SEC_OPTIONS["base_path"], config.PX_SEC_OPTIONS["secret_key"]):
+        return
+
+    dcos_utils.create_dcos_security_group(config.PX_SEC_OPTIONS["group_name"])
```
Why do you need group and user to be added to the group?
Can you not just create the user and give permissions to it?

jitupawar22 replied:
Noops! group is not needed, removing it.

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -190,6 +185,72 @@ def test_vol_delete():
+# Test dcos sectres prepare operations
+@pytest.mark.sanity
+def test_dcos_secrets_prepare():
+    dcos_utils.install_enterprises_cli()
```
This does not have to be a test. Could be just a function to setup secrets.
Rather test_dcos_secrets_present / asserting it is created should be just a part of it.

Remove the test_px_secrets_login test. Rather we should update the portworx service with secrets enabled and try creating a encrypted volume. So basically just one test.

jitupawar22 replied:
I kept it separate tests in order to speed up debugging in case of failures. 
Combining test_dcos_secrets_prepare and test_dcos_secrets_present.

Manually tried upgrading px framework/service with enabling secrets, but found that portworx-0-install does not start. there was no error related to offers. Also px service logs does not have any errors. the portworx-0-install node keeps failing continuously.  
I am keeping the test_px_secrets_login and test_create_encrypted_px_volume as it is until we have successful encrypted volume creation through px service update.

piyush-nimbalkar replied:
We should not modify our tests for that. If it is failing we should try to fix it. It is good if the test can catch that. Let's figure out first why it is was failing.

jitupawar22 replied:
sounds good

Commit:
Made changes as per review comments
Modifies pytest SDK for removing large unuseful logs and introduced retry for couple of items.
Did tunning of test run by introducing sleep at multiple places.

Commit:
Made sectrets test changes such that pxctl secrets login is not required.
The px service update by enabling secrets is sufficient.
Removed secrets tests from test_sanity and added it in test_updste.

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -93,37 +96,83 @@ def portworx_service(service_account):
+    # Restart portworx service on all px nodes
+    pod_count, pod_list = px_utils.get_px_pod_list()
+    if pod_count <= 0:
+        log.info("PORTWORX: Can't proceed with volume creation, Pod count is: {}".format(pod_count))
```
change the error message "Can't proceed with volume creation" is incorrect.

Commit:
Changing log message to appropriate one.
